distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/checks/EntitlementChecksService.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.runtime.checks;
+
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.checks.EntitlementProvider;
+import org.elasticsearch.entitlement.runtime.api.Entitlement;
+import org.elasticsearch.entitlement.runtime.api.FlagEntitlement;
+import org.elasticsearch.entitlement.runtime.api.NotEntitledException;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Collections.newSetFromMap;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isActive;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isFrozen;
+
+/**
+ * Implementation of the {@link EntitlementChecks} interface.
+ * The trampoline module loads this object via SPI.
+ */
+public class EntitlementChecksService implements EntitlementChecks {
+    /**
+     * @return the same instance of {@link EntitlementChecksService} returned by {@link EntitlementProvider}.
+     */
+    public static EntitlementChecksService get() {
+        return (EntitlementChecksService) EntitlementProvider.checks();
+    }
+
+    private final Set<Module> entitledToExit = newSetFromMap(new ConcurrentHashMap<>());
```

Comment:
We can't have a separate collection per entitlement. I think we need to have some kind of map of `entitlement key` -> `List<Module>` so it's all stored in the same structure

Reply:
Why?

Reply:
Oh also an earlier iteration of this was a `Map<Module, Set<FlagEntitlement>>`, if you like that better.

Reply:
Then this class will be full of member fields for each thing we want to do `entitledToExit`, `entitledToReadFile`, `entitledToWriteFile`, `entitledToUseReflection`... and adding a new type would mean a new field. Each type of entitlement is not part of the _structure_ of the class, but part of the _functionality_. The structure should not change with the functionality.

Reply:
Ok I see. Yeah I share your concern.

I could see it ending up in a small handful of data structures perhaps, with one for all "flag entitlements" (likely 80%+ of the entitlements) and then one for each kind of structured entitlement (say, files, networking, etc.). Each of these will need its own code for the entitlement checking logic anyway, so there's little to be saved by combining them into one data structure if they only need to be downcast to do the check anyway.

Reply:
`Elasticsearch` is not a public class. I used `ElasticsearchProcess` but now I've run into a problem where this is hard to unit test, because my tests are not in the same module as `ElasticsearchProcess`.

Reply:
For this PR I can simply hard-forbid `System.exit`. The unit test can disable that using `EntitlementInternals` so JUnit can shut down.